Basic Authentication Support

teromene opened the pull request:
Move configuration in its own class in order to reduce the verbosity of index.php

Added authentication mechanism using HTTP auth.

Maybe I should also move the configuration checks (ini and extension) to the new  class ?

teromene committed:
Move configuration in its own class in order to reduce the verbosity of index.php
Added authentication mechanism using HTTP auth.

logmanoriginal commented:
It's simple, it works... I like it :grinning: 

Who knew it was that easy :man_shrugging: 

For single-user environments this is certainly the easiest solution. Maybe in the future we can extend on it if the need arises (matter of time).

Splitting the configuration part into a separate class is also a good idea. index.php was getting to the point of being hard to read and comprehend. ~Maybe we can split off some more of the features into separate files in the future.~

Sorry, I should really learn to read properly :man_facepalming: 

>Maybe I should also move the configuration checks (ini and extension) to the new class ?

Yes, this is a good idea!

logmanoriginal commented on config.default.ini.php:
```diff
@@ -24,4 +24,9 @@
+[authentication]
+enable_authentication = false
+username =
+password =
```
I suppose `enable_authentication` should be renamed to `enable`, as `[authentication]` is redundant.

logmanoriginal replied:
Some explanation for each parameter would also be nice. Here is my suggestion:

```
[authentication]

; Enables authentication for all requests to this RSS-Bridge instance.
;
; Warning: You'll have to upgrade existing feeds after enabling this option!
;
; true  = enabled
; false = disabled (default)
enable = false

; The username for authentication. Insert this name when prompted for login.
username = ""

; The password for authentication. Insert this password when prompted for login.
; Use a strong password to prevent others from guessing your login!
password = ""
```

logmanoriginal commented on lib/Authentication.php:
```diff
@@ -0,0 +1,27 @@
+	
+	public static function showPromptIfNeeded() {
+		
+		if(Configuration::$config['authentication']['enable_authentication'] == true) {
```
`==` should be `===` for best practices

logmanoriginal commented on lib/Authentication.php:
```diff
@@ -0,0 +1,27 @@
+	public static function verifyPrompt() {
+		
+		if(isset($_SERVER['PHP_AUTH_USER']) && isset($_SERVER['PHP_AUTH_PW'])) {
+			return (Configuration::$config['authentication']['username'] == $_SERVER['PHP_AUTH_USER'] && Configuration::$config['authentication']['password'] == $_SERVER['PHP_AUTH_PW']);
```
Same as above (`===`)

logmanoriginal replied:
Do you think it makes sense to write a little comment to error_log? That way the number of failed attempts can be "tracked":

`error_log('Failed authentication attempt!')`

Of course this could be much more elaborate by including the IP, username and password of the client...

logmanoriginal commented on lib/Configuration.php:
```diff
@@ -0,0 +1,59 @@
+<?php
```
I do have mixed feeling about the syntax of `Configuration::$config['...']['...']`. It looks odd to me, although technically it is fine. I'd feel more comfortable with how other classes are implemented, using functions to access data:

```
class Configuration {
  // function to access configuration parameters (returns mixed data)
  public function get($parameter) { ... }

  ...
}

// index.php
$proxy_url = $config->get('proxy.url');
```

Of course all the loading could/should be done in the constructor

How do you see it?

logmanoriginal reviewed (changes requested):
Again, great work :+1: 

I decided to give feedback on a few things you might want to consider before merging. Let me know what you think on these points!

teromene committed:
Added a method to get the config parameters.
Removed the installation checks from the index page.
Log all failed authentication attemps.

teromene commented:
Modified to take your comments in account.

logmanoriginal commented:
Awesome, thanks!

I suppose this PR can be merged then...
